crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,373 @@
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_sk_X509_add1_cert(STACK_OF(X509) *sk, X509 *cert,
+                               int not_duplicate, int prepend)
+{
+    if (not_duplicate) {
+        /* not using sk_X509_set_cmp_func() and sk_X509_find()
+           because this re-orders the certs on the stack */
+        int i;
+        for (i = 0; i < sk_X509_num(sk); i++)
+            if (X509_cmp(sk_X509_value(sk, i), cert) == 0)
+                return 1;
+    }
+    if (!sk_X509_insert(sk, cert, prepend ? 0 : -1))
+        return 0;
+    return X509_up_ref(cert);
+}
+
+/*-
+ * Append list of certificates from 'certs' to given list,
+ * optionally only if not self-signed and
+ * optionally only if not already contained.
+ * The certs parameter may be NULL.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_sk_X509_add1_certs(STACK_OF(X509) *sk, const STACK_OF(X509) *certs,
+                                int no_self_signed, int no_duplicates)
+{
+    int i;
+
+    if (sk == NULL)
+        return 0;
+
+    if (certs == NULL)
+        return 1;
+    for (i = 0; i < sk_X509_num(certs); i++) {
+        X509 *cert = sk_X509_value(certs, i);
+        if (!no_self_signed || X509_check_issued(cert, cert) != X509_V_OK) {
+            if (!OSSL_CMP_sk_X509_add1_cert(sk, cert, no_duplicates, 0))
+                return 0;
+        }
+    }
+    return 1;
+}
+
+/*
+ * Add all or self-signed certificates from the given stack to given store.
+ * certs parameter may be NULL.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_X509_STORE_add1_certs(X509_STORE *store, STACK_OF(X509) *certs,
+                                   int only_self_signed)
+{
+    int i;
+
+    if (store == NULL)
+        return 0;
+
+    if (certs == NULL)
+        return 1;
+    for (i = 0; i < sk_X509_num(certs); i++) {
+        X509 *cert = sk_X509_value(certs, i);
+        if (!only_self_signed || X509_check_issued(cert, cert) == X509_V_OK)
+            if (!X509_STORE_add_cert(store, cert)) /* ups cert ref counter */
+                return 0;
+    }
+    return 1;
+}
+
+/*
+ * Retrieves a copy of all certificates in the given store.
+ * returns NULL on error
+ */
+STACK_OF(X509) *OSSL_CMP_X509_STORE_get1_certs(X509_STORE *store)
+{
+    int i;
+    STACK_OF(X509) *sk;
+    STACK_OF(X509_OBJECT) *objs;
+
+    if (store == NULL)
+        return NULL;
+    if ((sk = sk_X509_new_null()) == NULL)
+        return NULL;
+    objs = X509_STORE_get0_objects(store);
+    for (i = 0; i < sk_X509_OBJECT_num(objs); i++) {
+        X509 *cert = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(objs, i));
+        if (cert != NULL) {
+            if (!sk_X509_push(sk, cert)) {
+                sk_X509_pop_free(sk, X509_free);
+                return NULL;
+            }
+            X509_up_ref(cert);
+        }
+    }
+    return sk;
+}
+
+X509_EXTENSIONS *OSSL_CMP_X509_EXTENSIONS_dup(const X509_EXTENSIONS *extin)
```

Comment:
Good point that `OSSL_CMP_X509_EXTENSIONS_dup()` is strongly related to `sk_X509_EXTENSION_deep_copy` (which I did not know of so far). I've checked and found that the latter does not cope well with the case that its input is `NULL` (which can happen in our applications), so I've made  `OSSL_CMP_X509_EXTENSIONS_dup()` handle this case and otherwise call `sk_X509_EXTENSION_deep_copy`  with the respective dup and free functions.